shared/src/pool_collector.rs

```diff
@@ -0,0 +1,34 @@
+use crate::amm_pair_provider::AmmPairProvider;
+use crate::pool_fetching::{Pool, PoolFetcher, PoolFetching};
+use crate::Web3;
+use model::TokenPair;
+use std::collections::HashSet;
+use std::sync::Arc;
+
+pub struct PoolCollector {
+    pub pool_fetchers: Vec<PoolFetcher>,
+}
+
+impl PoolCollector {
+    pub fn new(pair_providers: Vec<Arc<dyn AmmPairProvider>>, web3: Web3) -> Self {
```

Comment:
Here I would pass in a list of PoolFetchers directly. ~This way we can reuse the `FilteredPoolFetcher<CachedPoolFetcher<PoolFetcher>>` layered component we already create in main.~

Actually, I see now how those are layered on top. I think we might need a separate trait interface when we want to generalize from just fetching `Pools` to fetching arbitrary `BaselineSolvable` objects, but we can cross this bridge when we get there. 

Yet, to leave this component very abstract (it aggregates different `impl PoolFetching` together), we could pass a `Vec<Arc<dyn PoolFetching>>`?

Reply:
Right, if we are going to do that then there really is no need for this new method since its only trying to reuse the web3 for `AmmPairProviders`. In that case I will just remove the new method so not to favour one implementation of PoolFetcher over another.